src/services/prediction.py

```diff
@@ -63,23 +63,33 @@ def get_rewards_per_epoch(
 
     @staticmethod
     def _group_events_by_transaction_hash(event_type_1: list[EventData], event_type_2: list[EventData]):
+        event_type_1_dict = {}
+
+        for event in event_type_1:
+            event_type_1_dict[event['transactionHash']] = event
+
+        if len(event_type_1_dict.keys()) != len(event_type_1):
+            raise ValueError('Events are inconsistent: some events from event_type_1 has same transactionHash.')
+
         result_event_data = []
 
-        for event_1 in event_type_1:
-            for event_2 in event_type_2:
-                if event_2['transactionHash'] == event_1['transactionHash']:
-                    result_event_data.append({
-                        **event_1['args'],
-                        **event_2['args'],
-                    })
-                    break
+        for event_2 in event_type_2:
+            tx_hash = event_2['transactionHash']
 
-        if len(event_type_1) == len(event_type_2) == len(result_event_data):
-            return result_event_data
+            event_1 = event_type_1_dict.pop(event_2['transactionHash'], None)
 
-        raise ValueError(
-            f"Events are inconsistent: {len(event_type_1)=}, {len(event_type_2)=}, {len(result_event_data)=}"
-        )
+            if not event_1:
+                raise ValueError(f'Events are inconsistent: no events from type 1 with {tx_hash=}.')
```

Comment:
suggesting to make this func less common and add some business logic specific to avoid confusing exceptions like this
in general we'd have no idea what the event type 1 or 2 here if they will be raised

Reply:
and we have literally one call of this func